### DISCUSSION TITLE
Full file system access ?

### DISCUSSION THREAD

**Discussion by erikpa1** [tauri-apps/tauri#7122] [Q&A]
Hi,

I would like to ask for, example ? maybe ? working codes snippet ? (no answer of type add "**" to your tauri config, like where ?? there is milion of settings)

I have hybrid 3D application, created with three js. App is multiplatform Desktop/Web. The app structure is based on "projects" where one folder is one project, with its assets, databases, files (images, videos, point clouds, fbx / glb files,...) One project can have 30 GB.

The projects can be located anywhere on PC, user has structure like:

C:\
  - Program Files\
  - ProjectData\
  
- D:\
  - Documents\
  - Projects\
  - Music\
  - ProjectData\
 
- E:\
  - Videos\
  - Games\
  - ProjectData\
Because I use a lot of external libraries, often they works like following

new three.FileLoader("/api/path/file.json")
new GltfLoader().load("/api/models/model.gltf")
So as you can see, I can work only with file paths synchronously (no async await), some libraries use for example folder + they compute file path like new PanoramaLoader().load(".api/pano/"), which hide "chunk1.png" "chunk2.png"...

I need to configure tauri, to be able load this files anywhere. No dialog picking, nothing similar, when user,

just when I write:

`GltfLoader().load("D:/Work/MyProject/preview.png")`

 it has to work.

I know tauri has some security standard, but this is basic workflow which app like Autocad, Autodesk, Blender, Unity, Unreal Engine uses, Visual Studio,... uses, and I want use it.

**Comment by FabianLars** [comment]
1) To load these files you'll need to use the asset: protocol via https://tauri.app/v1/api/js/tauri/#convertfilesrc - adding the os path into load() won't work like this.
2) Here's an example config: https://github.com/tauri-apps/tauri/blob/b02fc90f450ff9e9d8a35ee55dc1beced4957869/examples/api/src-tauri/tauri.conf.json#L112 just replace the array that's in there with `["**"]` to get your desired behavior.
2.5) If you also want to use tauri's fs apis like writeBinaryFile you'll also have to change the [fs scope](https://github.com/tauri-apps/tauri/blob/b02fc90f450ff9e9d8a35ee55dc1beced4957869/examples/api/src-tauri/tauri.conf.json#L90) accordingly.

Disclaimer for other people in case they stumble upon this issue: We heavily discourage this, and except recommend using tauri's dialog api, which will expand the scope automatically for selected files, or use rust to enable specific paths at runtime.

**Comment by erikpa1** [comment]
@FabianLars 
It's working, thank you for your time :)

**Comment by TatuLaras** [comment]
Does the solution still work as of 2024? 

My allowlist is as follows:
```json
...
            "protocol": {
                "asset": true,
                "assetScope": ["**"]
            },
...
```

And when I try to use `convertFileSrc` I still get the `asset protocol not configured to allow the path`.

**Reply by erikpa1** [reply]
@TatuLaras sorry for late response check this repo but it's not 2024 it tauri version "1.2.1": 

https://github.com/erikpa1/TurtleEngineFrontend/blob/main/src-tauri/tauri.conf.json

**Comment by maxbodin** [comment]
Is there a way to allow access to all the files, with tauri v2 ?

**Reply by Blaithnaid** [reply]
also looking for this, did you figure anything out?

**Reply by erikpa1** [reply]
@FabianLars  Examples for full availabilities of filesystem in tauri v2 are really required,.... could be really nice to boost documentation with examples...  like I understand some security rules Tauri has, but I think that there will be more applications requiring full filesystem access than than those who want fraction of it... some of us do real commercial software ,... A lot of questions is answered on discord missing the official documentation...

**Reply by FabianLars** [reply]
> some of us do real commercial software

Hm yeah, i'll give you that. Everyone knows that those products typically don't care about security at all (why choose a security-first framework like tauri then though?)

We're still talking about glob patterns and i hope that this is also mentioned here and there in the docs. `**` stopped working like in v1 because we follow the glob pattern standard more closely now. This means that you're probably looking for `**/*` now.

P.S. Everyone can contribute to our documentation ;) With how small our team currently is we have to rely on contributions.

**Reply by erikpa1** [reply]
@FabianLars  Yeah, I would like to contribute in doc, but I have not required level of  knowledge of tauri :(.

Also if you have time maybe I really would like to ask about "security philosophy" you are talking. Because most the applications I see on my desktop, are applications requiring full access. I have for example:

Obs studio (desktop recording) - user want to select where should it store files (and work with them later)
HandBrake (video editing software) - also user can store files anywhere(and work with them later)
Oracle Virtual Box - - also user can store files anywhere(and work with them later)
Unity - also user can store files anywhere(and work with them later)
Unreal Engine - also user can store files anywhere(and work with them later)
Blender - also user can store files anywhere(and work with them later)
Software for working with ebooks - also user can store files anywhere(and work with them later)
Photoshop - also user can store files anywhere(and work with them later)
Adobe Premier - also user can store files anywhere(and work with them later)
Visual studio code - also user can store files anywhere(and work with them later)
Servez - also user can store files anywhere(and work with them later)
MS Word - also user can store files anywhere(and work with them later)
MS Excell - also user can store files anywhere(and work with them later)
MS Project- also user can store files anywhere(and work with them later)
Drawio - also user can store files anywhere(and work with them later)

--------
Postman - ok, second case I can use only app data 
Age of Empires - ok, second case I can use only app data
OpenVPN gui -  ok another one

Let's sum it up... more application in my PC have requirement of having global access than those who don't... Why choose tauri - multiplatform....  Probably only security here about which we are talking is probably dependency injection in repos.... And again we are discussing about javascript injections, now I download rust package (probably injected with malware) and now my security is totally gone.

**Reply by FabianLars** [reply]
Huii that's a big list. Makes me realize that we do a very bad job explaining the dynamic nature of the scopes.

Whatever you define in the config is just the start, typically something you always need (like a location for app settings). Apologies for that.

Our dialog plugin for example allows endusers to choose a file- or directory path and will then add the chosen path to the scope automatically. To have that persist across app restarts you can use the persisted-scope plugin.
Same for file drag and drop btw (unless you disable it to be able to use the webviews native drag drop api).

If you need more control you can even extend the scope in rust at runtime via something like `app.fs_scope().allow_directory(path)`

If you (can) use all of that, there really aren't that many use-cases where `**/*` is required (personal preference aside)

**Reply by erikpa1** [reply]
@FabianLars Ok you have an point about things you talk about, it's true that I need mostly folder of file  where "myproject.turtle" (turtle is my own project format) was picked by user to save - this is true. I need the whole folder because files are uploaded in folder of project root.

Let's take unreal engine or unity as an example to have real world example. Basically what I develop has the same principle:

1. user opens launcher

2. user creates an project and select target destination - he can choose in dialog menu or he can paste path:
![image](https://github.com/user-attachments/assets/5da0d24b-d6df-42bc-92e6-2af8e47446c5)

4. now created project are in the list bellow:

![image](https://github.com/user-attachments/assets/5db8c346-25ec-49e3-ba4c-24cf2da6abc3)

5. User downloaded new version of my app and he needs to update project (run code made for refactor), and he needs to iterate all projects he see there (common situation in some industry apps - you change files format encodings compression,....).

And now there is a problem this thread is about - from documentation I have no idea how to do it. What I saw in documentation was only that I have put some permissions on some folders like Videos, AppData,... so my first idea coming on my mind - I need to somehow access full filesystem. I have no idea how to read the folder created by user in my app, and most crucial thing, I have no idea how to fetch data from that folder (yeah I know there is convertSrc,... but also this knowledge comes from discord not documentation)

I also tried to opened the Dialog in v2 tried to find anything about `app.fs_scope().allow_directory(path)` and I found nothing,... nothing about the case we discussed.

**Reply by erikpa1** [reply]
But to not just blame, and bring something positive, soon  I will share some application we developed in my company in multiplatform environment (android windows), it's nice 3D application based on three for playing 360 photos :)

**Reply by FabianLars** [reply]
Step 2 sounds exactly like the use case for the dialog api (for the Folder icon) and the rust api (for pasted paths). Both are not theoretical suggestions but approaches actually used in production apps so if something wasn't working for you it's something that may be worth to look into.

> I also tried to opened the Dialog in v2

Just so we're on the same page, i'm talking about https://v2.tauri.app/plugin/dialog/ not whatever the browser has built-in.

> tried to find anything about app.fs_scope().allow_directory(path) and I found nothing,... 

hm yeah that's probably not on the main website and only part of the rust docs eg https://docs.rs/tauri-plugin-fs/latest/tauri_plugin_fs/trait.FsExt.html#tymethod.fs_scope / https://docs.rs/tauri/2.4.1/tauri/scope/fs/struct.Scope.html#method.allow_directory (sadly most of our crates are poorly documented, even compared to our "standard")

**Reply by erikpa1** [reply]
I would never had idea to find it in and understand this doc https://docs.rs/tauri-plugin-fs/latest/tauri_plugin_fs/trait.FsExt.html#tymethod.fs_scope  .... :D :D  I think that more than documentation we need code snippets,...

**Reply by pburrows** [reply]
> Whatever you define in the config is just the start, typically something you always need (like a location for app settings). Apologies for that.

> Our dialog plugin for example allows endusers to choose a file- or directory path and will then add the chosen path to the scope automatically. To have that persist across app restarts you can use the persisted-scope plugin.

Definitely need to spell this out in the docs! I would not have realized that based on what is there now.
